chore: #191 Add documentation for launch Datamaintain (#192)
4sh/datamaintain: docs/as-dependency.md

</dependency>

```

## Use Datamaintain in your code

> You can take a look to the java sample for see an example of how to create and start datamaintain.

Reviewer: ```suggestion
You can take a look at the [java sample](../samples) for an example of how to create and start Datamaintain.
```

> For start a Datamaintain execution in your code, you need to create a `Datamaintain` class. 

Reviewer: ```suggestion
To start a Datamaintain execution in your code, you need to create a `Datamaintain` instance. 
```

The `Datamaintain` class is the core class.

> Two functions are available :

Reviewer: ```suggestion
Two functions are available:
```

> * `updateDatabase` : search and execute the db scripts. Same has the `update-db` command on the cli.

Reviewer: ```suggestion
* `updateDatabase`: search and execute the db scripts. Same as the `update-db` command on the cli.
```

Author: ```suggestion
* `updateDatabase`: search and execute the db scripts. Same has the `update-db` command on the cli.
```

> * `listExecutedScripts` : return a `Sequence` with all scripts known by Datamaintain. Same has the `list` command on the cli.

Reviewer: ```suggestion
* `listExecutedScripts`: return a `Sequence` with all scripts known by Datamaintain. Same as the `list` command on the cli.
```

Author: ```suggestion
* `listExecutedScripts`: return a `Sequence` with all scripts known by Datamaintain. Same as the `list` command on the cli.
```

> Datamaintain's constructor take a `DatamaintainDriverConfig`, there is two `Config objects`:

Reviewer: ```suggestion
Datamaintain's constructor takes a `DatamaintainDriverConfig`. There are two `Config objects`:
```

* A `DatamaintainDriverConfig` implementation (MongoDriverConfig, JdbcDriverConfig, ...) containing the database configuration
* `DatamaintainConfig` contains all datamaintain configuration, including the `DatamaintainDriverConfig`

You can create `Config objects` with properties, constructor or builder.

### Create Datamaintain Config Object
#### Properties
You can set all Datamaintain configuration in a property file: see [properties documentation](configuration.md).

> From your code you can load the property file for create the `Config objects`, each `Config` class has a `buildConfig` method for that:

Reviewer: ```suggestion
From your code you can load the property file to create the `Config objects`. Each `Config` class has a `buildConfig` method for that:
```

```properties
scan.path=/scripts
scan.identifier.regex=(.*?)_.*
db.uri=mongodb://localhost:27017/datamaintain
```
```kotlin
val properties = Properties()
properties.load(MyClass::class.java.getResourceAsStream("datamaintain.properties"))

val driverConfig: DatamaintainDriverConfig = DatamaintainConfig.buildConfig(properties)
val config = DatamaintainConfig.buildConfig(driverConfig, properties)
```

#### Kotlin
You can create `Config objects` directly from constructor in pure kotlin:

```kotlin
val driverConfig: DatamaintainDriverConfig = MongoDriverConfig(uri = "mongodb://localhost:27017/datamaintain")
val config = DatamaintainConfig(
    path = Paths.get("/scripts"),
    identifierRegex = "(.*?)_.*",
    driverConfig = driverConfig
)
```

#### Java

> Like Kotlin, you can create object directly from constructor, but in Java you cannot use Kotlin default parameter. 

Reviewer: ```suggestion
Like Kotlin, you can create object directly from constructor, but in Java you cannot use Kotlin default parameters. 
```

> It is difficult to create and maintain a Java code that use the constructor. 

Reviewer: ```suggestion
It is difficult to create and maintain a Java code that uses the constructor. 
```

> For Java user, Datamaintain provided a `Builder` class in each `Config` class:

Reviewer: ```suggestion
For Java user, Datamaintain provides a `Builder` class in each `Config` class:
```

```java
public class MyClass {
    final DatamaintainDriverConfig driverConfig = new MongoDriverConfig.Builder()
            .withUri("mongodb://localhost:27017/datamaintain-sample-java-mongo")
            .build();

    final DatamaintainConfig config = new DatamaintainConfig.Builder()
            .withPath(Paths.get("/scripts"))
            .withIdentifierRegex(new Regex("(.*?)_.*"))
            .withDriverConfig(driverConfig)
            .build();
}
```

### Create and use the Datamaintain object

> You can create the `Datamaintain` object from a `DatamaintainConfig` and then call the list or update method :

Reviewer: ```suggestion
You can create the `Datamaintain` object from a `DatamaintainConfig` and then call the list or update method:
```

```kotlin
val driverConfig: DatamaintainDriverConfig = ...
val config = ...

Datamaintain(config).updateDatabase()
```

### Exception

> All Datamaintain exception use a custom exception. There is 2 exception to known :

Reviewer: ```suggestion
All Datamaintain exceptions use a custom exception. There are 2 exceptions to know:
```

#### DatamaintainBaseException

> All Datamaintain exceptions extends this class. In addtion to the message field, a resolutionMessage field 

Reviewer: ```suggestion
All Datamaintain exceptions extend this class. In addition to the message field, a resolutionMessage field 
```

can help to resolve.

> ### DatamaintainException

Reviewer: ```suggestion
#### DatamaintainException
```

> This class extends `DatamaintainBaseException`. It is thrown if an exception occur in a step code.

Reviewer: ```suggestion
This class extends `DatamaintainBaseException`. It is thrown if an exception occurs in a step code.
```

> This class add two fields :

Reviewer: ```suggestion
This class adds two fields:
```

> * `step` that indicate the step that throw the exception (SCAN, FILTER, ...) 

Reviewer: ```suggestion
* `step` that indicates the step that throw the exception (SCAN, FILTER, ...) 
```
